File: GO_Backend/internal/crypto/hmac.go
package crypto

import (
	"crypto/hmac"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
)

// SignMessage creates an HMAC-SHA256 signature of the given message body.
func SignMessage(body []byte, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write(body)
	return hex.EncodeToString(mac.Sum(nil))
}

// VerifySignature checks that the provided signature matches the HMAC-SHA256
// of the message body using the given secret.
func VerifySignature(body []byte, signature, secret string) bool {
	expected := SignMessage(body, secret)
	return hmac.Equal([]byte(expected), []byte(signature))
}

// SignJSON marshals the payload to JSON and produces an HMAC-SHA256 signature.
func SignJSON(payload interface{}, secret string) ([]byte, string, error) {
	data, err := json.Marshal(payload)
	if err != nil {
		return nil, "", err
	}
	sig := SignMessage(data, secret)
	return data, sig, nil
}

// BlindIndex generates an HMAC-SHA256 blind index for searchable encrypted fields.
func BlindIndex(value, secret string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(value))
	return hex.EncodeToString(mac.Sum(nil))
}

File: backend/src/utils/hmac.js
const crypto = require('crypto');

const HMAC_SECRET = process.env.INTERNAL_HMAC_SECRET || 'dev-hmac-secret';

/**
 * Sign a message payload with HMAC-SHA256.
 * Used for internal Node↔Go service-to-service request signing.
 */
const signMessage = (payload) => {
  const body = typeof payload === 'string' ? payload : JSON.stringify(payload);
  return crypto.createHmac('sha256', HMAC_SECRET).update(body).digest('hex');
};

/**
 * Verify an HMAC-SHA256 signature against a payload.
 */
const verifySignature = (payload, signature) => {
  const expected = signMessage(payload);
  return crypto.timingSafeEqual(Buffer.from(expected, 'hex'), Buffer.from(signature, 'hex'));
};

/**
 * Generate a blind index using HMAC-SHA256 for searchable encrypted fields.
 */
const blindIndex = (value) => {
  const key = process.env.BLIND_INDEX_KEY || '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef';
  return crypto.createHmac('sha256', Buffer.from(key, 'hex'))
    .update(value.toLowerCase().trim())
    .digest('hex');
};

module.exports = { signMessage, verifySignature, blindIndex };
